scripts/validate_docs_snippet_line_numbers.py

```diff
@@ -0,0 +1,202 @@
+
+    Args:
+        directory: The dir to recursively search through.
+
+    Returns:
+        A map between Python files and the docs that reference them.
+    """
+    docusaurus_refs: Dict[str, List[DocusaurusRef]] = defaultdict(list)
+    for file in glob.glob(f"{directory}/**/*.md", recursive=True):
+        logger.debug("Checking %s for Docusaurus links", file)
+        _collect_docusaurus_refs(file=file, reference_map=docusaurus_refs)
+
+    return docusaurus_refs
+
+
+def _collect_docusaurus_refs(
+    file: str, reference_map: Dict[str, List[DocusaurusRef]]
+) -> None:
+    with open(file) as f:
+        contents = f.read()
+
+    references: List[Tuple[str, str, str]] = DOCUSAURUS_REGEX.findall(contents)
+
+    for reference in references:
+        docusaurus_ref = DocusaurusRef.from_regex_parsed_values(
+            file=file, values=reference
+        )
+        reference_map[docusaurus_ref.py_file].append(docusaurus_ref)
+        logger.debug(
+            "Create association between %s and the docs that use it",
+            docusaurus_ref.py_file,
+        )
+
+
+def evaluate_snippet_validity(
+    docusaurus_refs: Dict[str, List[DocusaurusRef]]
+) -> List[DocusaurusRef]:
+    """Given a series of references, reviews the referenced files and ensures that
+    opening and closing tags directly surround the snippet.
+
+    Ex: ```python file=...#L2
+
+        1 <snippet>
+        2
+        3 context = DataContext()
+        4 </snippet>
+
+    Note that the tags do not need to be immediately before and after the content - as long
+    as the lines in between are whitespace, the algorithm will not flag it as an issue.
+
+    Args:
+        docusaurus_refs: A map between Python files and the docs that reference them.
+
+    Returns:
+        A list of broken references.
+    """
+    all_broken_refs: List[DocusaurusRef] = []
+    for docs_file, ref_list in docusaurus_refs.items():
+        logger.debug(
+            "Checking %s for snippet validity",
+            docs_file,
+        )
+        file_broken_refs: List[DocusaurusRef] = _evaluate_snippet_validity(
+            docs_file, ref_list
+        )
+        all_broken_refs.extend(file_broken_refs)
+
+    return all_broken_refs
+
+
+def _evaluate_snippet_validity(
+    file: str, ref_list: List[DocusaurusRef]
+) -> List[DocusaurusRef]:
+    with open(file) as f:
+        file_contents: List[str] = f.readlines()
+
+    broken_refs: List[DocusaurusRef] = []
+    for ref in ref_list:
+        if not _validate_docusaurus_ref(ref, file_contents):
+            broken_refs.append(ref)
+
+    return broken_refs
+
+
+def _validate_docusaurus_ref(
+    docusaurus_ref: DocusaurusRef, file_contents: List[str]
+) -> bool:
+    start_line, end_line = docusaurus_ref.line_numbers
+    if start_line > len(file_contents) or end_line > len(file_contents):
+        logger.warning(
+            "Invalid snippet due to out-of-bounds reference: %s", docusaurus_ref
+        )
+        return False
```

Comment:
We've got a few of these - although Docusaurus is okay with out-of-bounds errors, I'd like to ensure we're accurately referencing lines so our docs are accurate and correct.